[New feature] Allowing the `ListView` slivers to have different extents while still having scrolling performance

xu-baolin opened the pull request:
Fixes https://github.com/flutter/flutter/issues/113431

Currently we only support specifying all slivers to have the same extent.
This patch introduces an `itemExtentBuilder` property for `ListView`, allowing the slivers to have different extents while still having scrolling performance, especially when the scroll position changes drastically(such as scrolling by the scrollbar or controller.jumpTo()).

@Piinks Hi, Any thoughts about this?  :)

xu-baolin committed:
Adding the itemExtentCallback property to the ListView allows slivers to have different extents with high layout performance.

xu-baolin committed:
++

Piinks commented on packages/flutter/lib/src/rendering/sliver.dart:
```diff
@@ -16,6 +16,9 @@ import 'viewport_offset.dart';
 // The RenderSliver base class and its helper types.
 
+/// Called to get the item extent by the index of item.
+typedef ItemExtentGetter = double Function(int index);
```
What do you think of passing an index plus a little bit more information about the viewport?

My original thought for this was to see what folks thought of the TableSpanExtent API in the TableView (https://github.com/flutter/packages/pull/4536) and maybe move that into the framework.
This section of the design document talks about it (links direct to that section): https://docs.google.com/document/d/15ecTZE1g3WeswLGFWrnEgMP6SyL6jDRdxOgPsczOcV0/edit?resourcekey=0-yNd_qFhiPjz6z2TgezWc0A#heading=h.a8uex5ucyiso

It defines multiple ways for developers to set an extent, based on the viewport, or the preceding scroll extent, or as the result of an operation, etc. 

I figured if it worked well for folks (we did already test it in a usability study), we could use it here as well, making it a super class like `ChildScrollExtent` or some such - names are hard. 😆

xu-baolin replied:
```What do you think of passing an index plus a little bit more information about the viewport?```

What do you think of exposing the `SliverConstraints` directly or just some of the key properity of viewport?

Piinks replied:
Hmmm, that's a good idea! It won't communicate the information about how much the SliverList has laid out already though, since the SliverConstraints relate to the whole sliver and not its contents. What parts of SliverConstraints would be relevant to the user when building an item of the list? Definitely the viewportExtent I think. Maybe we can wrap those parts up in an object with additional info like how much extent has been laid out.

xu-baolin replied:
I have wraped some layout relation properity into `SliverLayoutDimensions` and exposed it to the callback.

Will finish the test cases ASAP.

Piinks reviewed (commented):
This looks very exciting! Thank you for putting this together!!
I am have been meaning to get to something similar, just have not had the opportunity yet. I thought this type of approach would resolve https://github.com/flutter/flutter/issues/52207

xu-baolin committed:
review feedback

xu-baolin committed:
++

xu-baolin committed:
add tests

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
++

Piinks commented on packages/flutter/lib/src/rendering/sliver.dart:
```diff
@@ -16,6 +16,67 @@ import 'viewport_offset.dart';
+/// Called to get the item extent by the index of item.
+typedef ItemExtentGetter = double Function(int index, SliverLayoutDimensions dimensions);
+
+/// Holds the relation dimensions of the [RenderSliver] during layout.
```
```suggestion
/// Relates the dimensions of the [RenderSliver] during layout.
///
/// Used by...
```

Can add a reference to the item extent callback from sliver list / list view?

Piinks commented on packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart:
```diff
@@ -170,28 +200,60 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
-      maxExtent: itemExtent,
+    _currentLayoutDimensions = SliverLayoutDimensions(
+        scrollOffset: constraints.scrollOffset,
+        precedingScrollExtent: constraints.precedingScrollExtent,
```
Ah ok, it looks like this is not updated as each child is laid out. I think users will expect it to account for the children in the SliverList that came before it in addition to any slivers that came before it. Maybe SliverLayoutDimensions can have a copyWith method, and we can add each child's extent to preceding extent as we lay them out and call on the item extent getter?

xu-baolin replied:
>I think users will expect it to account for the children in the SliverList that came before it in addition to any slivers that came before it.
If I understand correctly, you mean when call `itemExtentBuilder(1000, dimentions)`, `dimentions.precedingScrollExtent` should equal `constraints.precedingScrollExtent + sumExtentsOf(0~999)`?

If that's the case, developers can calculate this value themselves, right? Or the `dimentions.scrollOffset` is more meaningful? Otherwise introducing a new `precedingScrollExtent` would be confused with `SliverConstraints.precedingScrollExtent`. 

Additional, during a frame, `itemExtentBuilder` will be called many times, and the called item does not mean that it will be loaded/layout by the lazy loading mechanism. I mean, when scrolling to the end, all items‘s callback will be called.

Piinks replied:
Oh you are right, and make a really good point about opportunities for confusion there. The user could also keep track of the extents as the builder is called. Thanks for explaining!

Piinks replied:
Also! I just thought, maybe it should be documented somewhere that the itemExtentBuilder could be called multiple times so users are aware 👍

xu-baolin replied:
Added some docs to `flutter.widgets.list_view.itemExtentBuilder`

```dart
  ///
  /// This will be called multiple times during the layout phase of a frame to find
  /// the items that should be loaded by the lazy loading process.
  ///

```

Piinks reviewed (commented):
So excited about this. 😁  Thanks for the update!

xu-baolin committed:
review feedback

xu-baolin committed:
++

xu-baolin committed:
++

Piinks commented on packages/flutter/lib/src/rendering/sliver.dart:
```diff
@@ -16,6 +16,69 @@ import 'viewport_offset.dart';
 // CORE TYPES FOR SLIVERS
 // The RenderSliver base class and its helper types.
 
+/// Called to get the item extent by the index of item.
```
Can you add references here to where it is used?

Piinks commented on packages/flutter/lib/src/widgets/sliver_explicit_extent_list.dart:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  @override
+  double get itemExtent => double.nan;
```
Could this be nullable? Or maybe we can add an assertion as the body of this getter that says this should not be used?

xu-baolin replied:
Yes, I have change it to nullable and add a assertion at the beginning of `RenderSliverFixedExtentBoxAdaptor.performLayout()`
```dart
    assert((itemExtent != null && itemExtentBuilder == null) ||
        (itemExtent == null && itemExtentBuilder != null));
    assert(itemExtentBuilder != null || (itemExtent!.isFinite && itemExtent! >= 0));
```

Piinks commented on packages/flutter/lib/src/widgets/sliver_explicit_extent_list.dart:
```diff
@@ -0,0 +1,137 @@
+  /// A sliver that places multiple box children in a linear array along the main
+  /// axis.
+  ///
+  /// [SliverFixedExtentList] places its children in a linear array along the main
```
Is this referring to the wrong one?

xu-baolin replied:
fixed

Piinks commented on packages/flutter/lib/src/widgets/sliver_explicit_extent_list.dart:
```diff
@@ -0,0 +1,137 @@
+  /// A sliver that places multiple box children in a linear array along the main
+  /// axis.
+  ///
+  /// [SliverFixedExtentList] places its children in a linear array along the main
```
Same here

xu-baolin replied:
fixed

Piinks commented on packages/flutter/lib/src/widgets/sliver_explicit_extent_list.dart:
```diff
@@ -0,0 +1,137 @@
+///    extent in the main axis.
+///  * [SliverFillViewport], which sizes its children based on the
+///    size of the viewport, regardless of what else is in the scroll view.
+class SliverExplicitExtentList extends SliverMultiBoxAdaptorWidget {
```
In general, does this need to be it's own class? Would it be possible to use SliverFixedExtentList for both cases? I am curious for your thinking first, no need to change the code right now. As a developers, I wonder, what is the difference between Fixed and Explicit?

xu-baolin replied:
I have considered extending SliverFixedExtentList to support variable extent. 
But I'm worried that this will break the developer's understanding of SliverFixedExtentList.
Maybe we need a better name to tell them the difference between the two. 
It's really hard to name it, any suggestions? 😄

xu-baolin replied:
That's a great name : )

Piinks commented on packages/flutter/lib/src/rendering/sliver_fixed_extent_list.dart:
```diff
@@ -46,6 +47,9 @@ abstract class RenderSliverFixedExtentBoxAdaptor extends RenderSliverMultiBoxAda
   double get itemExtent;
 
+  /// The main-axis extent builder of each item.
+  ItemExtentGetter? get itemExtentBuilder => null;
```
What if when null, this builder just returns itemExtent for every index by default? Maybe it could simplify the code below with `if (itemExtentBuilder == null)`?

xu-baolin replied:
I have added some docs for this.
```dart
  /// The main-axis extent builder of each item.
  ///
  /// If this is non-null, the [itemExtent] will be ignored.
  /// If this is null, the [itemExtent] must be positive finite.
  ItemExtentBuilder? get itemExtentBuilder => null;
```

Piinks reviewed (commented):
I think this is nearly there. Thank you very much for putting this together, and for your patience. I found the issue requesting this and added it to the description.

xu-baolin committed:
review feedback

xu-baolin committed:
++

xu-baolin committed:
++

xu-baolin committed:
rename the new widget and RO

xu-baolin committed:
rename the file name

xu-baolin committed:
++

xu-baolin committed:
++

GeylanKalafMohe commented:
Hello please approve

Piinks reviewed (approved):
LGTM! I am really excited for folks to have this, thank you so much for your work on this! 🎉

oddko commented:
Thank you so much for your great work !

xu-baolin commented:
@Piinks Hi, Thank you very much for reviewing.
Hope you guys enjoyed your trip to Shanghai last week : )

Oh, by the way, I have recently been trying to introduce a caching mechanism to ListView to reduce unnecessary layout operations to further optimize its scrolling performance. I hope to find a better solution.

发件人: Kate Lovett ***@***.***>
发送时间: 2023年9月12日 8:29
收件人: flutter/flutter ***@***.***>
抄送: 徐宝林(Ava) ***@***.***>; Author ***@***.***>
主题: Re: [flutter/flutter] [New feature] Allowing the `ListView` slivers to have different extents while still having scrolling performance (PR #131393)


@Piinks approved this pull request.

LGTM! I am really excited for folks to have this, thank you so much for your work on this! 🎉

—
Reply to this email directly, view it on GitHub<https://github.com/flutter/flutter/pull/131393#pullrequestreview-1621053588>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOR66GBPUMFRG2IH6KLUPMTXZ6UDJANCNFSM6AAAAAA2ZV6CVM>.
You are receiving this because you authored the thread.Message ID: ***@***.******@***.***>>

________________________________
OPPO

本电子邮件及其附件含有OPPO公司的保密信息，仅限于邮件指明的收件人（包含个人及群组）使用。禁止任何人在未经授权的情况下以任何形式使用。如果您错收了本邮件，切勿传播、分发、复制、印刷或使用本邮件之任何部分或其所载之任何内容，并请立即以电子邮件通知发件人并删除本邮件及其附件。
网络通讯固有缺陷可能导致邮件被截留、修改、丢失、破坏或包含计算机病毒等不安全情况，OPPO对此类错误或遗漏而引致之任何损失概不承担责任并保留与本邮件相关之一切权利。
除非明确说明，本邮件及其附件无意作为在任何国家或地区之要约、招揽或承诺，亦无意作为任何交易或合同之正式确认。 发件人、其所属机构或所属机构之关联机构或任何上述机构之股东、董事、高级管理人员、员工或其他任何人（以下称“发件人”或“OPPO”）不因本邮件之误送而放弃其所享之任何权利，亦不对因故意或过失使用该等信息而引发或可能引发的损失承担任何责任。
文化差异披露：因全球文化差异影响，单纯以YES\OK或其他简单词汇的回复并不构成发件人对任何交易或合同之正式确认或接受，请与发件人再次确认以获得明确书面意见。发件人不对任何受文化差异影响而导致故意或错误使用该等信息所造成的任何直接或间接损害承担责任。
This e-mail and its attachments contain confidential information from OPPO, which is intended only for the person or entity whose address is listed above. Any use of the information contained herein in any way (including, but not limited to, total or partial disclosure, reproduction, or dissemination) by persons other than the intended recipient(s) is prohibited. If you are not the intended recipient, please do not read, copy, distribute, or use this information. If you have received this transmission in error, please notify the sender immediately by reply e-mail and then delete this message.
Electronic communications may contain computer viruses or other defects inherently, may not be accurately and/or timely transmitted to other systems, or may be intercepted, modified ,delayed, deleted or interfered. OPPO shall not be liable for any damages that arise or may arise from such matter and reserves all rights in connection with the email.
Unless expressly stated, this e-mail and its attachments are provided without any warranty, acceptance or promise of any kind in any country or region, nor constitute a formal confirmation or acceptance of any transaction or contract. The sender, together with its affiliates or any shareholder, director, officer, employee or any other person of any such institution (hereinafter referred to as "sender" or "OPPO") does not waive any rights and shall not be liable for any damages that arise or may arise from the intentional or negligent use of such information.
Cultural Differences Disclosure: Due to global cultural differences, any reply with only YES\OK or other simple words does not constitute any confirmation or acceptance of any transaction or contract, please confirm with the sender again to ensure clear opinion in written form. The sender shall not be responsible for any direct or indirect damages resulting from the intentional or misuse of such information.

Piinks commented:
> Hope you guys enjoyed your trip to Shanghai last week : )

Absolutely! Meeting each other was a highlight of the trip! 🎉 

> Oh, by the way, I have recently been trying to introduce a caching mechanism to ListView to reduce unnecessary layout operations to further optimize its scrolling performance. I hope to find a better solution.

This is really interesting, I would be happy to hear more if you are interested in pursuing it. I think in some cases caching could work, we could see what it looks like to give developers the option. Not sure. :)
